Update color texture build to reduce collisions, and fix collision handling

Czaki opened the pull request:
# Description

During work on #6179 I found that the mechanism for avoiding collision does not work correctly (see tests). I also found that the shader code do not have a stop condition. 

In this PR, I also add a dynamic calculation of the hash table size. I also decide to use prime numbers as the size of the table to reduce collision probability.

Czaki committed:
some code clean

Czaki committed:
Fix colision in keys

Czaki committed:
add dynamic calculation of texture size

Czaki committed:
prevent from infinite loop

Czaki committed:
improve tests

Czaki committed:
speedup tests

Czaki committed:
check texture maximum size

Czaki commented on napari/_vispy/layers/labels.py:
```diff
@@ -164,11 +173,46 @@ def hash2d_set(key, value, keys, values, empty_val=0):
+            f' Requested size is {shape[0]}x{shape[1]}, but maximum supported'
+            f' size is {MAX_TEXTURE_SIZE[0]}x{MAX_TEXTURE_SIZE[1]}'
+        )
+    return shape
```
@brisvag what did you think about this implementation 
(based on https://github.com/napari/napari/pull/6179#issuecomment-1688051007)

brisvag replied:
I think this makes sense. Why exactlu are you using a global variable here?

Czaki replied:
Performance. But now I check the code, and I need to fix this, as it is the max size for 2d and 3d not the max size of the texture dimension.

Czaki replied:
Ok. I have changed to limit texture size to use the size that will work in both 2d and 3d mode.

brisvag replied:
> Performance. But now I check the code, and I need to fix this, as it is the max size for 2d and 3d not the max size of the texture dimension.

Ah true!

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -68,8 +73,12 @@
         t = t + 1;
+        if (t - initial_t > LUT_shape.x * LUT_shape.y) {
+            return vec4(0);
+        }
```
Good catch, I remember my gpu hanging a few times when working on this, probably because of this.

Czaki committed:
limit to 3d maximum texture size

Czaki committed:
use minimum of possible texture size

Czaki committed:
Merge branch 'main' into bugfix/label_direct_coloring

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -195,17 +195,17 @@ def _get_shape_from_dict(color_dict):
     global MAX_TEXTURE_SIZE
     if MAX_TEXTURE_SIZE is None:
-        MAX_TEXTURE_SIZE = get_max_texture_sizes()
+        MAX_TEXTURE_SIZE = min(get_max_texture_sizes())
```
min? Wha tif one or the other is None? Why not just get the 2d (`[0]`)?

Czaki replied:
for example on my machine 2d is bigger than 3d:

```python 
from napari._vispy.utils.gl import get_max_texture_sizes
print(get_max_texture_sizes())
```

provides: `(32768, 16384)`

brisvag replied:
But we're using a 2D texture, so it doesn't matter what the allowed 3D shape is.

Czaki replied:
So 3D is for `sampler2D`?

brisvag replied:
I guess you mean `sampler3D`, in which case yes. Regardless whether we look at 3D or 2D labels, the lookup table for coloring is a 2D texture!

Czaki replied:
fixed

Czaki committed:
fix texture size

jni commented on napari/_vispy/layers/labels.py:
```diff
@@ -68,8 +73,12 @@
         t = t + 1;
+        if (t - initial_t > LUT_shape.x * LUT_shape.y) {
+            return vec4(0);
+        }
```
```suggestion
    float initial_t = t;
    float lut_size = LUT_shape.x * LUT_shape.y
    while ((abs(found - initial_t) > 1e-8) && (abs(found - empty) > 1e-8)) {
        t = t + 1;
        if (t - initial_t > lut_size) {
            return vec4(0);
        }
```

maybe a micro-optimisation but since that condition is there for *many* loop locations, I thought it might be worth doing?

brisvag replied:
I'm pretty sure this is easily optimized by the compiler, but I guess it doesn't hurt...

Czaki committed:
Update napari/_vispy/layers/labels.py

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

Czaki commented on napari/_vispy/layers/labels.py:
```diff
@@ -68,8 +73,13 @@
     // - otherwise, it's a hash collision: continue searching
-    while ((abs(found - t) > 1e-8) && (abs(found - empty) > 1e-8)) {
+    float initial_t = t;
+    float lut_size = LUT_shape.x * LUT_shape.y
```
There is one problem here. As float has precision up to `2**23`. Should I limit the number of labels to `2**23` in function to construct?

brisvag replied:
I would say so!

brisvag replied:
Actually, should we also change the inequality comparisons from using `1e-8` to `1e23`? Or is that unrelated? I don't remember this code well anymore ^^'

jni replied:
> Should I limit the number of labels to 2**23 in function to construct?

What? No! The issue with the float32 `2**23` is that labels *might* end up with the same color. But they might not! `2**25 + 100` and `2**25 + 1000` will fall in different bins:

```
In [21]: np.float32(2**25 + 1000) - np.float32(2**25 + 100)
Out[21]: 900.0
```

Occasional key overwriting is better than crashing! And this issue is already tracked by #6084.

Czaki replied:
I mean that limit option to provide more than `2**23` different labels.

brisvag replied:
Yeah the question is about total number of labels... though to be fair, I guess even then it's better to just have guaranteed duplicates than simply crashing :shrug:

jni replied:
Even for total number it doesn't make sense? labels `1` to `2**23` will all work fine together. Then you have every other label between `2**23` and `2**24`. And so on. So there are many more than ~8M labels that could be separately rendered.

Czaki replied:
I have updated the limitations. Could you check them? 
I also need to make PR to vispy (for future improvements)

Czaki replied:
> Even for total number it doesn't make sense? labels `1` to `2**23` will all work fine together. Then you have every other label between `2**23` and `2**24`. And so on. So there are many more than ~8M labels that could be separately rendered.

The limitation comes from the condition stop of the while loop. As I change to use int (`int32`) instead of float the problem is solved in another way.

jni replied:
> The limitation comes from the condition stop of the while loop. As I change to use int (int32) instead of float the problem is solved in another way.

OOOOooooooooOOOOH! I get it now! 😅 Clever. 😅

Czaki committed:
better limit number of possible labels

Czaki committed:
move one bellow

jni commented on napari/_vispy/layers/labels.py:
```diff
@@ -164,11 +174,46 @@ def hash2d_set(key, value, keys, values, empty_val=0):
+def _get_shape_from_dict(color_dict):
+    size = len(color_dict) * 4
+    # I think that hash table size should be at least 4 times
+    # bigger than the number of labels to avoid collisions
```
Should we have a MAX_LOAD_FACTOR constant, and maybe the tests should test the load factor rather than specific shapes? It would make it easier to run some experiments and revisit this decision in the future...

Czaki replied:
The load factor here is `0.25`. I'm not sure how to understand next part.

jni replied:
I mean it's nicer when the code refers to a named constant rather than `4`.

Additionally, the tests check that we return a specific shape. But that shape will change if we change the default load factor in the future. Instead, we should check that the load of the table is between 1/4 and 1/16. (ie between LOAD_FACTOR and LOAD_FACTOR / EXPANSION_FACTOR, which in this case is the same.)

Czaki replied:
I will think. It is quite easy to test if the shader is enough big, but I'm not sure how to simply test if it is the smallest possible shader.

jni commented on napari/_vispy/layers/labels.py:
```diff
@@ -164,11 +176,51 @@ def hash2d_set(key, value, keys, values, empty_val=0):
+        raise OverflowError(
+            f'Too many labels. GPU does not support textures of this size.'
+            f' Requested size is {shape[0]}x{shape[1]}, but maximum supported'
+            f' size is {MAX_TEXTURE_SIZE}x{MAX_TEXTURE_SIZE}'
```
I think maybe this should be a MemoryError? OverflowError is a subclass of ArithmeticError, which is not what happened here, but rather a failure to allocate a large enough texture?

Czaki replied:
I just use the same error as exists in code for this context. We could even implement one exception (`ToBigForShader` for example). What did you think?

jni replied:
I generally dislike custom exceptions because then you need to import something to handle them. Personally I would go with a MemoryError here.

Czaki replied:
Then, I prefer to create such an exception inheriting from an existing, less specific one. But I changed it to MemoryError.

Czaki committed:
fix problem with adding 1 to float

jni commented on napari/_vispy/layers/labels.py:
```diff
@@ -77,8 +77,8 @@
     while ((abs(found - initial_t) > 1e-8) && (abs(found - empty) > 1e-8)) {
-        t = t + 1;
         count = count + 1;
+        t = initial_t + float(count);
```
💯

jni reviewed (commented):
Thanks @Czaki! I think this is very close! I'll have another look tomorrow! 🙏

Czaki commented on napari/_vispy/layers/labels.py:
```diff
@@ -236,6 +288,7 @@ def _on_colormap_change(self, event=None):
                 interpolation='nearest',
             )
             self.node.shared_program['LUT_shape'] = key_texture.shape
+            self.node.shared_program['color_count'] = len(color_dict) + 1
```
Hm. I do not know what I should use here. Using `len(color_dict) + 1` will spedup stop searching for missing entries, but may cast some real big values to transparent if a collision happens.

jni replied:
How so? Suppose all the labels are 1...n, then one more label m such that m lands on 1. This means that it will have to scan through 1-n, then land on n+1. So `len(color_dict)+1` will be sufficient to not skip that possibility, right? In what other scenario will it be too small?

Czaki replied:
For keys above `2**23`. But I may be wrong.

jni replied:
Well, this is handled by a uint32 (nice improvement!), so I don't think it will be an issue? (you will run out of texture shape before you run out of precision / overflow.)

Czaki commented on napari/_vispy/layers/labels.py:
```diff
@@ -68,8 +74,14 @@
+    while ((abs(found - initial_t) > 1e-8) && (abs(found - empty) > 1e-8)) {
+        count = count + 1;
+        t = initial_t + float(count);
+        if (count >= color_count) {
+            return vec4(0);
+        }
```
If the count is above 0 and we hit an empty field then we could assume that color is not defined. Did I'm correct?

Czaki replied:
Yes. It is, and it is handled by while loop condition

Czaki committed:
deduplicate colormap creation

Czaki committed:
add collision check when set colors

Czaki committed:
add direct mode for coloring without collisions

Czaki committed:
proper handle casting to float32

Czaki committed:
fix mypy

Czaki committed:
fix coordinate types

Czaki committed:
Merge branch 'main' into bugfix/label_direct_coloring

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Czaki committed:
fix nan condition

Czaki committed:
test collision check

Czaki committed:
use MAX_LOAD_FACTOR

jni commented on napari/layers/labels/labels.py:
```diff
@@ -575,6 +586,25 @@ def color(self, color):
+            "Because of the cast to float, the following labels will be merged: "
+            + ", ".join([f"{l1} and {l2}" for l1, l2 in collision_list])
+        )
+        warnings.warn(warn_text, category=RuntimeWarning)
```
I think this should be in a different PR tbh... One issue is that the aliasing will happen *both* in direct color mode (where you are adding this check and it's roughly as fast as everything else so not a big deal) and in auto color mode (where engineering a fast solution will be more challenging, especially since new labels can be added at any time). So it feels like a halfway solution.

At any rate, I think both the docstring and the warning should link to #6084, so people can get more details about what's happening. Also, the labels don't get "merged", that sounds like you are messing with their data which is not what's happening.

Finally, I think that pairs are not the right way to present this — for large values, 8, 16, 32 and more labels will get aliased together. So I think they should be presented as lists rather than as pairs.

Anyway, some suggestions:

```suggestion
    def _validate_colors(self, labels: Iterable[int]):
        """Check whether any of the given labels will be aliased together.
        
        See https://github.com/napari/napari/issues/6084 for details.
        """
        aliases: Dict[float, List[int]] = defaultdict(list)
        labels_int = np.fromiter(labels, dtype=int)
        labels_texture = self._to_vispy_texture_dtype(labels_int)
        for integer, texture in zip(labels_int, labels_texture):
            aliases[texture].append(integer)
        aliased_list = [v for k, v in aliases.items() if len(v) > 1]
        if not aliased_list:
            return
        alias_strings = [
            f'Labels {lst[:-1]} will display as the same color as {lst[-1]};'
            for lst in aliased_list
        ]

        warn_text = trans._(
            "Because integer labels are cast to less-precise float for display, "
            "the following label sets will render as the same color:\n"
        ) + '\n'.join(alias_strings) + trans._(
            "See https://github.com/napari/napari/issues/6084 for details."
        )
        warnings.warn(warn_text, category=RuntimeWarning)
```

I forgot how to translate f-strings but that should also be translated.

In the interest of time I suggest this should all get moved to a separate PR because it deals with an orthogonal issue.

Czaki replied:
extracted to #6193

Czaki committed:
change exception to MemoryError

Czaki committed:
adjust test to handle changes of MAX_LOAD_FACTOR

Czaki committed:
Revert "add collision check when set colors"

This reverts commit daae57892fab7f6d49eeca0d3776921bfa6c8268.

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -98,6 +125,49 @@
 """
 
 
+direct_lookup_shade_without_collision = """
```
Should also be changed below.

```suggestion
direct_lookup_shader_without_collision = """
```

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -68,8 +89,14 @@
-        t = t + 1;
+    float initial_t = t;
+    int count = 0;
+    while ((abs(found - initial_t) > 1e-8) && (abs(found - empty) > 1e-8)) {
```
Sorry if this is a dumb question, I mostly forgot how this code works... But is it ok that we have `1e-8` epsilon in this comparison here, or should we go lower?

Czaki replied:
the difference between keys should be 1. So 1e-8 is really strict.

brisvag replied:
Ah right, this is just keys comparison, my bad!

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -120,12 +190,18 @@ def __init__(
+        else:
+            self.glsl_map = direct_lookup_shade_without_collision.replace(
+                '$use_selection', str(use_selection).lower()
+            ).replace('$selection', str(selection))
```
Two things I'm thinking of here:
1. these two shaders are basically identical except for the extra while loop for collision checking. Do we really need to maintain 2 separate shaders? I think it would be better to at least have a single shader with a bool uniform like `check_collision` which simply decides what to do.
2. I know this was already here before this PR, but we should probably use vispy's replacement logic instead of this hacky replace code. You should be able to do it with something like this:
```py
from vispy.visuals.shaders import Function

func = Function(direct_lookup_shader)
func['use_selection'] = use_selection
func['selection'] = selection
func['check_collision'] = collision
self.glsl_map = func
```

Czaki replied:
I do not know. I do not exactly know how shader compilation works, how it cut dead code. In general the code executed on GPU do not like `if` clauses. 

The code replacement is to allow compiler to optimize code during compilation time and to not execute redundant checks. So if it is possible, then replacement should be used instead of providing variable value.

brisvag replied:
As far as I know, this is only a problem if branching differs between fragments. If all fragments have the same branching (e.g: because of a bool uniform) then it should be compiled away.

As for the variables already used (`use_selection` and `selection`), vispy already replaces them inplace, except it has some extra checks to ensure everything is consistent, so the replacement there should be exactly identical in terms of final compilation result.

Czaki replied:
I have checked. Even if vispy documentation claims that you are right this vispy feature is not working :P

brisvag replied:
In what way it doesn't work? I have used it in other places before, so it should work in some way...

brisvag replied:
(but this is not blocking for this PR, if you prefer to move forward). ~Though I do think it would be best to merge the 2 shaders if possible.~

Czaki replied:
It declares a global variable but not do replace `$variable_name`. The shade compilation fails.

brisvag replied:
This works as intended with just the `Function` object:

```py
from vispy.visuals.shaders import Function

low_disc_lookup_shader = """
uniform sampler2D texture2D_LUT;

vec4 sample_label_color(float t) {
    float phi_mod = 0.6180339887498948482;  // phi - 1
    float value = 0.0;
    float margin = 1.0 / 256;

    if (t == 0) {
        return vec4(0);
    }

    if (($use_selection) && ($selection != t)) {
        return vec4(0);
    }

    value = mod((t * phi_mod + $seed), 1.0) * (1 - 2*margin) + margin;

    return texture2D(
        texture2D_LUT,
        vec2(0.0, clamp(value, 0.0, 1.0))
    );
}
"""

func = Function(low_disc_lookup_shader)
func['seed'] = 1
func['selection'] = 2
func['use_selection'] = True

print(func.compile())
```

Something must be broken in how the `glsl_map` is being used...

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -143,37 +219,87 @@ def hash2d_get(key, keys, values, empty_val=0):
+def _get_shape_from_dict(color_dict):
+    size = len(color_dict) / MAX_LOAD_FACTOR
+    # I think that hash table size should be at least 4 times
+    # bigger than the number of labels to avoid collisions
```
A comment explaining what load factor and prime num table are would be great here!

Czaki replied:
could you check current comment?

brisvag replied:
Ok, I have a few comment/questions:

If that was meant to be a docstring, it should be moved one line up.

I'm confused about the load factor. It's 0.25, so this makes sense:
```
    I think that hash table size should be at least four times
    bigger than the number of labels to avoid collisions
```
but this seems to go contradict it:
```
    MAX_LOAD_FACTOR is used to determined the maximum number of labels
    as factor of hash table size. Literature propose to use table at
    least 1.3 times bigger than number of labels. But it is for classical
    approach where hash table entry could store a list of values. In our case
    we have to store only one value per entry.
    So I think that we could use bigger load factor.
```
So I would expect a load factor > 1.3? you mean that it's basically 4 since you divide by 0.25? Either way, it's not super clear to me why you think a bigger value than literature here is a better idea.


I'm also not sure I get this:
```
    I decide to use it as a hash table size because the next size of hash
    table will be around two times bigger.
```
Do you mean that each subsequent element is roughly twice the previous? And that's good why?

Czaki replied:
> If that was meant to be a docstring, it should be moved one line up.

No. I do not have idea to make it docstring, but maybe I should? 




> So I would expect a load factor > 1.3? you mean that it's basically 4 since you divide by 0.25? Either way, it's not super clear to me why you think a bigger value than literature here is a better idea.

The problem here is cost of collisions and cost of branched calculation on GPU. When you have factor 1.3 (0,76) there are some collisions, but on CPU they are much cheaper as there is no worker group that work in synchronization. 

> Do you mean that each subsequent element is roughly twice the previous? And that's good why?

First shader will be 61x61, second 61x127, third 127x127 etc. there is a need to made some decision. Doubling size looks good for me (except first one the shader will be maximum 8 times bigger than labels number).

brisvag replied:
> No. I do not have idea to make it docstring, but maybe I should?

Maybe yes, I think it's better.

> The problem here is cost of collisions and cost of branched calculation on GPU. When you have factor 1.3 (0,76) there are some collisions, but on CPU they are much cheaper as there is no worker group that work in synchronization.

Ok, so basically compared to the "ideal" factor, we prefer over-using memory to gain a bit fewer collisions, because we prefer to maximize rendering performance instead of memory footprint?

> First shader will be 61x61, second 61x127, third 127x127 etc. there is a need to made some decision. Doubling size looks good for me (except first one the shader will be maximum 8 times bigger than labels number).

Ok... I guess this is a reasonable heurisitic, I don't have an intuition for what's best here :P

Czaki replied:
I reimplement it to reduce collision. Could you check now?

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -143,37 +219,87 @@ def hash2d_get(key, keys, values, empty_val=0):
+def build_textures_from_dict(color_dict, empty_val=0, shape=None):
+    if len(color_dict) > 2**31 - 2:
+        raise OverflowError(
+            'Too many labels. Maximum supported number of labels is 2^31-2'
```
```suggestion
            f'Too many labels ({len(color_dict)}). Maximum supported number of labels is 2^31-2'
```

brisvag reviewed (commented):
Generally good! I left a couple comments that I think should be addressed before merging.

Czaki commented on napari/_vispy/layers/labels.py:
```diff
@@ -120,12 +190,18 @@ def __init__(
+                '$use_selection', str(use_selection).lower()
+            ).replace('$selection', str(selection))
+        else:
+            self.glsl_map = direct_lookup_shade_without_collision.replace(
```
```suggestion
            self.glsl_map = direct_lookup_shader_without_collision.replace(
```

Czaki committed:
Apply suggestions from code review

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

Czaki committed:
improve comment

Czaki committed:
improve calcualtion of prime table size

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -248,34 +252,51 @@ def hash2d_set(key: float, value, keys, values, empty_val=0) -> bool:
+    """
+    Check if there is a collision in the hashmap.
+    """
+    for fst_size, snd_size in product(
```
Love this!

```suggestion
def _get_shape_from_keys(keys, fst_dim, snd_dim):
    """
    Get the smallest hashmap size without collisions, if any.
    """
    for fst_size, snd_size in product(
```

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -248,34 +252,51 @@ def hash2d_set(key: float, value, keys, values, empty_val=0) -> bool:
+
+    We use PRIME_NUM_TABLE to get precomputed prime numbers.
+    We decided to use primes close to powers of two, as they
+    allowed to keep fill of has table between 0.125 to 0.25
```
```suggestion
    Get the shape of the 2D hashmap from the number of labels.
    The hardcoded shapes use prime numbers designed to avoid collisions.
    As the current collision resolution is non-linear, we decide to
    use hash table of size around four times bigger than the number
    of labels, instead of the classical 1.3 times bigger, to
    minimize the chance of collision in the shader at the cost
    of some video memory.

    We use PRIME_NUM_TABLE to get precomputed prime numbers.
    We decided to use primes close to powers of two, as they
    allowed to keep fill of hash table between 0.125 to 0.25
```

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -248,34 +252,51 @@ def hash2d_set(key: float, value, keys, values, empty_val=0) -> bool:
+        if res is None:
+            res = _get_shape_from_keys(keys, fst_dim, snd_dim + 1)
+        if res is None:
+            return PRIME_NUM_TABLE[fst_dim][0], PRIME_NUM_TABLE[snd_dim][0]
```
Still missing some comments here to follow this try/if block!

Czaki committed:
Apply suggestions from code review

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

Czaki committed:
improve descriptions

Czaki committed:
use one shader code

Czaki committed:
hardcode color count

Czaki committed:
reduce double emmision

Czaki committed:
fix type annotation

Czaki committed:
fix type annotation

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -63,11 +64,15 @@
+        if ($selection == t) {
+            texture2D(
+            texture2D_values,
+            vec2(0.0, 0.0)
```
missing a return or a var definition here? Not sure what the intention was.

Czaki replied:
missing return

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -95,23 +100,25 @@
+        while ((abs(found - initial_t) > 1e-8) && (abs(found - empty) > 1e-8)) {
+            count = count + 1;
+            t = initial_t + float(count);
+            if (count >= $color_count) {
```
nice!

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -297,7 +266,12 @@ def _get_shape_from_dict(color_dict):
     except IndexError:
-        return PRIME_NUM_TABLE[-1], PRIME_NUM_TABLE[-1]
+        # Index error means that we have too many labels to fit in 2**16.
+        if (max_size := PRIME_NUM_TABLE[-1][0] ** 2) <= len(color_dict):
```
You sure this is `<=` and not `<`?

Czaki replied:
changed

brisvag commented on napari/layers/labels/labels.py:
```diff
@@ -710,7 +710,7 @@ def color_mode(self, color_mode: Union[str, LabelColorMode]):
         self._selected_color = self.get_color(self.selected_label)
         self.events.color_mode()
-        self.events.colormap()
+        self.events.colormap()  # If remove this emitting, connect shader update to color_mode
```
Why this choice? It seems better to have more informative events, and just handle them the same downstream, instead of emitting the same event here.

Czaki replied:
Because the calculation hash map is not trivial (and not so fast), and I have not found another way to do this single. 

Also, real change is not connecting to the `color_mode` event in VispyLayer. And this comment is to reconnect in such a situation.

brisvag replied:
Yes I saw, that's what I mean. The change basically doesn't change how many times the shader is updated, because anyways you're connecting to `layer.events.colormap()` right?

Czaki replied:
Without this, after change `color_mode` the shader is updated twice, After it it is updated once.

Czaki committed:
use reall collison table

Czaki committed:
fix condition

Czaki committed:
fix typing

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -38,11 +64,19 @@
+            texture2D_values,
+            vec2(0.5, 0.5)
+        );
+        };
```
```suggestion
    if ($use_selection) {
        // just single-color texture is passed in this case
        if ($selection == t) {
            return texture2D(
                texture2D_values,
                vec2(0.5, 0.5)
            );
        };
```

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -68,19 +102,26 @@
+            }
+        }
+        // not found in collision table
+        return vec4(0);
```
When does this happen exactly? Or is this technically unreachable?

Czaki replied:
When data contains a label that is not covered by color dict, but hash calculated is the same as for the defined color so `found` key will be different from `t` but there will be no entry in the collision table for this.

brisvag replied:
ah ok, makes sense!

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -120,60 +161,220 @@ def __init__(
 
 
-def hash2d_get(key, keys, values, empty_val=0):
+def hash2d_get(key, keys, col_key) -> Tuple[Optional[Tuple[int, int]], bool]:
```
I thought it was like `column_key` at first

```suggestion
def hash2d_get(key, keys, collision_keys) -> Tuple[Optional[Tuple[int, int]], bool]:
```

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -120,60 +161,220 @@ def __init__(
+def hash2d_set(key: float, value, keys, values, empty_val=0) -> bool:
     """
     Set a value in the 2d hashmap, wrapping around to avoid collision.
     """
```
```suggestion
def hash2d_set(key: float, value, keys, values, empty_val=0) -> bool:
    """
    Set a value in the 2d hashmap if possible, returning True on collision.
    """
```

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -120,60 +161,220 @@ def __init__(
+
+
+def _get_shape_from_keys(
+    keys: np.ndarray, fst_dim: int, snd_dim: int
```
Would be best to rename variables here to be more explicit, like `second` instead of `snd`. What does `crd` stand for, actually?

Also, some comments or a docstring in this function would be great, I don't understand how those numbers and calculations are optimizing for low collisions, and it would be hard to maintain without asking you...

Czaki replied:
updated, could you check?

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -120,60 +161,220 @@ def __init__(
+
+def _build_collision_table(
+    dkt: Dict[float, Tuple[float, float, float]]
+) -> Tuple[np.ndarray, np.ndarray]:
```
```suggestion
def _build_collision_table(
    color_dict: Dict[float, Tuple[float, float, float]]
) -> Tuple[np.ndarray, np.ndarray]:
```

Czaki replied:
updated

brisvag commented on napari/_vispy/layers/labels.py:
```diff
@@ -221,10 +428,23 @@ def _on_colormap_change(self, event=None):
                 use_selection=colormap.use_selection,
                 selection=colormap.selection,
+                collision=collision,
+                color_count=len(color_dict) + 1,
```
why +1? background label?

Czaki replied:
removed

brisvag reviewed (commented):
I think I get the general idea: every time the label key set changes, we:
- check against every possible/reasonably sized table shapes to see how many collisions would happen
- pick one with no collisions if possible, otherwise minimize collisions
- construct a normal hashmap for anything non-colliding
- construct a second map for colliding labels, which is also sized appropriately

What I don't quite get is:
- how this second table makes things better (don't we have to deal with collisions there as well?)
- how all these magic numbers result in lower collisions (some links for future reference would be great)
- whether this is becoming too expensive. We might be entering benchmark territory here, because we don't want to make it super heavy on the CPU to simply add a new label if this just costs us some collision on the GPU and a relatively small hit on the FPS.

Czaki committed:
Apply suggestions from code review

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

Czaki committed:
add more docstring and basic tests

Czaki committed:
fix types

Czaki committed:
Merge branch 'main' into bugfix/label_direct_coloring

jni committed:
Change scope and add docstring to new fixture to clarify use

jni committed:
Add comment describing the prime_num_table

Czaki committed:
restore previous implementation of lookup function

Czaki committed:
Merge remote-tracking branch 'napari/main' into bugfix/label_direct_coloring

Czaki committed:
fix typing

jni committed:
Make _get_shape_from_keys docstring PEP-257 compliant

jni committed:
Update outdated _get_shape_keys_from_dict docstring

jni committed:
Simplify logic (no need for nonlocal IndexError)

Also, use MemoryError instead of OverflowError when we can’t allocate a big enough texture.

jni committed:
Remove unused _build_collision_table

jni committed:
Fix indentation error in docstring

jni committed:
Add comment clarifying “duplicate” dict keys

jni committed:
Fix black formatting due to LHS parentheses

jni reviewed (approved):
@Czaki I've made some simplifying changes, which I verified pass the tests. I also changed OverflowError -> MemoryError as previously discussed. Otherwise this is ready to go imho! 🙏
